images/encryption/blockcipher/blockcipher_siv.go

```diff
@@ -0,0 +1,219 @@
+	outsize   int64  // output size
+}
+
+// AESSIVEncryptor implments the CryptedDataReader for en- and decrypting input
+// data and calculating the hash over the data
+type AESSIVEncryptor interface {
+	CryptedDataReader
+}
+
+type aessivcryptor struct {
+	bc           *AESSIVLayerBlockCipher
+	outputReader io.Reader
+	digester     digest.Digester
+}
+
+// NewAESSIVLayerBlockCipher returns a new AES SIV block cipher of 256 or 512 bits
+func NewAESSIVLayerBlockCipher(bits int) (LayerBlockCipher, error) {
+	if bits != 256 && bits != 512 {
+		return nil, errors.New("AES SIV bit count not supported")
+	}
+	return &AESSIVLayerBlockCipher{bits: bits}, nil
+}
+
+func (r *aessivcryptor) Read(p []byte) (int, error) {
+	if r.bc.err != nil {
+		return 0, r.bc.err
+	}
+
+	for {
+		// return data if we have any
+		if r.bc.outbuffer != nil && r.bc.outoffset < len(r.bc.outbuffer) {
+			n := copy(p, r.bc.outbuffer[r.bc.outoffset:])
+			r.bc.outoffset += n
+			r.digester.Hash().Write(p[:n])
+
+			return n, nil
+		}
+		// no data and hit eof before?
+		if r.bc.eof {
+			return 0, io.EOF
+		}
+		// read new data; we expect to get r.bc.toread number of bytes
+		// for anything less we assume it's EOF
+		numbytes := 0
+		for numbytes < r.bc.toread {
+			var n int
+			n, r.bc.err = r.bc.reader.ReadAt(r.bc.inbuffer[numbytes:r.bc.toread-numbytes], r.bc.inoffset+int64(numbytes))
+			numbytes += n
+			if r.bc.err != nil {
+				if r.bc.err == io.EOF {
+					r.bc.eof = true
+					r.bc.err = nil
+					break
+				} else {
+					return 0, r.bc.err
+				}
+			}
+			if n == 0 {
+				break
+			}
+		}
+		if numbytes < r.bc.toread {
+			r.bc.eof = true
+		}
+
+		r.bc.inoffset += int64(numbytes)
+
+		// transform the data
+		if r.bc.encryptor != nil {
+			r.bc.outbuffer = r.bc.encryptor.Seal(nil, r.bc.inbuffer[:numbytes], []byte(""), r.bc.eof)
+		} else {
+			r.bc.outbuffer, r.bc.err = r.bc.decryptor.Open(nil, r.bc.inbuffer[:numbytes], []byte(""), r.bc.eof)
+			if r.bc.err != nil {
+				return 0, r.bc.err
+			}
+		}
+		// let reader start from beginning of buffer
+		r.bc.outoffset = 0
+		r.bc.outsize += int64(len(r.bc.outbuffer))
+	}
+}
+
+// Size returns the Size of the encrypted data
+func (r *aessivcryptor) Size() int64 {
+	return r.bc.outsize
+}
+
+func (r *aessivcryptor) Digest() digest.Digest {
+	return r.digester.Digest()
+}
+
+func (r *aessivcryptor) Hash() hash.Hash {
+	return r.digester.Hash()
+}
+
+// init initializes an instance
+func (bc *AESSIVLayerBlockCipher) init(encrypt bool, reader io.ReaderAt, opt LayerBlockCipherOptions) (LayerBlockCipherOptions, error) {
+	var (
+		err   error
+		nonce []byte
+		se    miscreant.StreamEncryptor
+	)
+
+	bc.reader = reader
+
+	key := opt.SymmetricKey
+	if len(key) != bc.bits/8 {
+		return LayerBlockCipherOptions{}, fmt.Errorf("invalid key length of %d bytes; need %d bytes", len(key), bc.bits/8)
+	}
+
+	nonceStr := opt.CipherOptions["nonce"]
+	if nonceStr != "" {
+		nonce, err = base64.StdEncoding.DecodeString(nonceStr)
```

Comment:
Why is the nonce base64 encoded at this point? It seems weird to me to make `CipherOptions` a string map and put the decoding everywhere it is used, rather than putting the keys in explicitly, in which case Go will convert byte types from base64 automatically.

Reply:
`CipherOptions` is a map[string]string but the Nonce is a []byte. It looks like that map only carries the Nonce at the moment, so it could be converted to map[string][]byte and we probably wouldn't have to do the conversion then...

Reply:
I converted the map to map[string][]byte, which resolve the issue: https://github.com/containerd/containerd/pull/3001/commits/d647440bfde890556230e6eff430cb56d57ec1c3